dummy file location change

Isan-Rivkin opened the pull request:
Closes #6471 

**Solution**
- Create `dummy` file in `_lakefs/dummy` 
- Maintains backward compatibility (repositories created prior to this version will be checked to with `dummy` at the root namespace) 
- Will break if downgraded and repositories created in this version.
- Side Effect: Removed support for migration from lakeFS version < v0.50. It's already outdated and otherwise the dummy file raises more questions. (i.e what happens if skipping accessibility test when repo created on Azure).

arielshaqed commented on docs/howto/deploy/upgrade.md:
````diff
@@ -136,86 +136,3 @@ To do so, add the following to your configuration (relevant **only** for 0.30.0)
 cataloger:
   type: rocks
 ```
-
````
Can we instead add a note along these lines?

```suggestion
If you are using a version before 0.50.0, you must first perform the [previous upgrade to that version](https://docs.lakefs.io/v0.50/reference/upgrade.html#data-migration-for-version-v0500). {: note: .note-warning }

Isan-Rivkin replied:
👍

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
-		dummyData = "this is dummy data - created by lakeFS in order to check accessibility"
-	)
+	dummyKey := fmt.Sprintf("%s/dummy", c.Config.Committed.BlockStoragePrefix)
+	const dummyData = "this is dummy data - created by lakeFS in order to check accessibility"
```
If we're making changes...
```suggestion
	const dummyData = "this is dummy data - created by lakeFS to check accessibility"
```

Isan-Rivkin replied:
breaking backward compatibility with transient storage but we can live with that 😃

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
 		StorageNamespace: storageNamespace,
 		IdentifierType:   block.IdentifierTypeRelative,
-		Identifier:       dummyKey,
+		Identifier:       "dummy",
```
Suggest keeping this as a key `dummyLegacyKey` or somesuch.  If you do that you probably don't need the comment at l.1632.

Isan-Rivkin replied:
Changed implementation please look

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
-	if _, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
-		return fmt.Errorf("found lakeFS objects in the storage namespace(%s): %w",
+
+	if _, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
```
YIKES!  As Mitch Hedberg would say,
> _I used to leak a file descriptor.  I still do.  But I used to, too.

Consider using `.Exists()` (which doesn't test readability), or just closing (which probably also tests readability):
```suggestion
	if exists, err := c.BlockAdapter.Exists(ctx, objLegacyLocation, objLen); err == nil {
```
or
```suggestion
	if s, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
		s.Close()
```

Isan-Rivkin replied:
Good 👁️ you have!

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
+		Identifier:       dummyKey,
+	}
+
+	if _, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
```
Obviously the same applies here, too.

Isan-Rivkin replied:
Changed implementation please look

arielshaqed commented on pkg/api/controller.go:
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
+		Identifier:       dummyKey,
+	}
+
+	if _, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
```
Bug?
```suggestion
	if _, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
```

arielshaqed replied:
Probably merits a test in controller_test.go.  The important part is to test that re-creating another repo at the same location fails; less important is to verify that creating another repo at the same location as a "legacy" dummy repo also fails.

Isan-Rivkin replied:
isn't it what [use same storage namespace twice](https://github.com/treeverse/lakeFS/blob/master/pkg/api/controller_test.go#L203) test does?

arielshaqed reviewed (changes requested):
Neat!  Some minor but important requests.

arielshaqed commented:
> * Side Effect: Removed support for migration from lakeFS version < v0.50. It's already outdated and otherwise the dummy file raises more questions. (i.e what happens if skipping accessibility test when repo created on Azure).

This is important for the CHANGELOG.  Can you add a

Isan-Rivkin commented on Makefile:
```diff
@@ -165,6 +165,10 @@ build: gen docs ## Download dependencies and build the default binary
 	$(GOBUILD) -o $(LAKEFS_BINARY_NAME) -ldflags $(LD_FLAGS) -v ./cmd/$(LAKEFS_BINARY_NAME)
 	$(GOBUILD) -o $(LAKECTL_BINARY_NAME) -ldflags $(LD_FLAGS) -v ./cmd/$(LAKECTL_BINARY_NAME)
 
+# TODO(isan) delete before merge
```
👍

Isan-Rivkin commented:
@arielshaqed please check again, I addressed all of your great comments + changed the implementation. 

Regarding "Legacy": 
Before recent commit: I did legacy check if `dummy` or `_lakefs/dummy` exists, now I changed it to only check `_lakefs/dummy`. 
The reason is that when clients will use minimal IAM permissions they will not be able to `Get` object `dummy`, they will get permission deny and that creates weird behavior questions in lakeFS - that can be addressed by adding more configuration but I'd rather not keeping this backward compatible for simplicity - @ozkatz please verify that's what we discussed.

arielshaqed commented on pkg/config/config.go:
```diff
@@ -228,9 +228,10 @@ type Config struct {
-		Local                  *struct {
+		Type                        string  `mapstructure:"type" validate:"required"`
+		DefaultNamespacePrefix      *string `mapstructure:"default_namespace_prefix"`
+		EnsureRootNamespaceRWAccess bool    `mapstructure:"ensure_root_namespace_rw_access"`
```
Yay, naming!  AFAICT this only causes read-access to be verified, and not even that (if you have just new repos).

Consider
```suggestion
		ReadableRootNamespace bool    `mapstructure:"readable_root_namespace"`
```

Also arguably this is better a part of Graveler than of Blockstore: it has no influence on blockstore behaviour.

arielshaqed replied:
@nopcoder probably can say.

arielshaqed commented on pkg/config/defaults.go:
```diff
@@ -60,6 +60,7 @@ func setDefaults(cfgType string) {
 	viper.SetDefault("auth.remote_authenticator.default_user_group", "Viewers")
 	viper.SetDefault("auth.remote_authenticator.request_timeout", 10*time.Second)
 
+	viper.SetDefault("blockstore.ensure_root_namespace_rw_access", true)
```
And also update docs/reference/configuration.md, please :-/

Isan-Rivkin replied:
done

arielshaqed reviewed (approved):
Thanks!

Approved to save time, but I think we should improve the config option name, and we **must** document it.

Isan-Rivkin commented on pkg/api/controller.go:
```diff
@@ -1626,20 +1626,21 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
-			storageNamespace, ErrStorageNamespaceInUse)
+
+	if s, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
+		defer s.Close()
```
👍
